### PR TITLE
Tags propagation: added incompatible flag

### LINKED ISSUE #7766: Tags to execution requirements propagation in Starlark
Every target in Bazel supports specifying a `tags` attribute:
```
cc_binary(
  ...
  tags = ["local"]
)
```

While tags can be used for a variety of use cases (i.e. query), Bazel also defines a set of tags that have special meaning for how it executes an action. For example, `no-sandbox` tells Bazel to run this action without a sandboxing strategy and `no-remote` tells it to not run the action on a remote execution system.

The issue is that a user sets `tags` on a target and not on an action. So a rule implementation needs to intentionally pass the tags forward to every action it creates (as an execution requirement).

```
def _impl(ctx):
    ctx.actions.run(
       ...
       execution_requirements = {tag:"" for tag in ctx.attr.tags.keys()},
    )

foo = rule(
    implementation = _impl,
    attrs = {
      ...
    },
)
```

If a rule doesn't add `execution_requirements = ctx.attr.tags.keys()` or similar to each action it creates then a user tagging a target as `no-remote` will have no effect for execution.

We need to answer the following questions:
* Is it desirable for Starlark rules to by default implicitly pass along a defined set of tags (i.e. the ones supported by Bazel) as an execution requirement to every action?
* Would enabling this by default be a backwards incompatible change?
* What are the semantics when a rule also defines their own execution requirements?

cc: @laurentlb @cparsons @nlopezgi @agoulti @jmmv

### PR DESCRIPTION
**Pull request by ishikhman**
reverted rollback of tags propagation: https://github.com/bazelbuild/bazel/commit/29eecb56c5f0f26b4751ec9eb79954412fb2991c

Tags declared on targets are not propagated to actions and therefore are not taken into consideration by bazel. This causes some issues, for instance, target marked with a tag 'no-remote' will still be executed remotely.
As it was agreed in the design doc (see #7766 for a link), set of tags to be propagated to actions as a first iteration.
This change is responsible for that first step for the Starlark Rules.

RELNOTES: tags 'no-remote', 'no-cache', 'no-remote-cache', 'no-remote-exec', 'no-sandbox' are propagated now to the actions from targets when '--ncompatible_allow_tags_propagation' flag set to true. See #8830.

Closes #7766
Related to #8830

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/packages/StarlarkSemanticsOptions.java [hunk 1]
```diff
@@ -176,6 +176,20 @@
...
+      documentationCategory = OptionDocumentationCategory.STARLARK_SEMANTICS,
+      effectTags = {OptionEffectTag.BUILD_FILE_SEMANTICS},
+      metadataTags = {
+          OptionMetadataTag.INCOMPATIBLE_CHANGE,
+          OptionMetadataTag.TRIGGERED_BY_ALL_INCOMPATIBLE_CHANGES
+      },
+      help =
+          "If set to true, tags will be propagated from a target to the actions' execution requirements;"
+              + " otherwise tags are not propagated")
+  public boolean incompatibleAllowTagsPropagation;
```
Comment (buchgr) [inline src/main/java/com/google/devtools/build/lib/packages/StarlarkSemanticsOptions.java]: link to issue?

**Comment by buchgr** [comment]
LGTM.

File: src/main/java/com/google/devtools/build/lib/packages/StarlarkSemanticsOptions.java (hunk 1 above)
Comment (ishikhman) [reply src/main/java/com/google/devtools/build/lib/packages/StarlarkSemanticsOptions.java]: done, thanks!

File: src/test/java/com/google/devtools/build/lib/packages/TargetUtilsTest.java [hunk 2]
```diff
@@ -96,4 +98,177 @@ public void testExecutionInfo() throws Exception {
     execInfo = TargetUtils.getExecutionInfo(tag1b);
     assertThat(execInfo).containsExactly("local", "", "cpu:4", "");
   }
+
+  @Test
+  public void testExecutionInfo_withPrefixSupports() throws Exception {
+    scratch.file(
+        "tests/BUILD",
+        "sh_binary(name = 'with-prefix-supports', srcs=['sh.sh'], tags=['supports-workers',"
+            + " 'supports-whatever', 'my-tag'])");
```
Comment (hlopko) [inline src/test/java/com/google/devtools/build/lib/packages/TargetUtilsTest.java]: Nit: scratch.file accepts varargs for the content, putting each string on a separate line, so this works:
scratch.file(
  "tests/BUILD",
  "sh_binary(",
  "  name = 'asdf',",
  "  srcs = [...],",
...)

**Pull request closed without merging**